WordPress/src/main/java/org/wordpress/android/ui/reader/viewmodels/tagsfeed/ReaderTagsFeedViewModel.kt

```diff
@@ -98,74 +92,98 @@ class ReaderTagsFeedViewModel @Inject constructor(
+                    tag = tag,
+                    posts = posts,
+                    onTagClick = ::onTagClick,
+                    onSiteClick = ::onSiteClick,
+                    onPostCardClick = ::onPostCardClick,
+                    onPostLikeClick = ::onPostLikeClick,
+                    onPostMoreMenuClick = ::onPostMoreMenuClick,
+                    onItemEnteredView = ::onItemEnteredView,
                 )
             } else {
-                updatedLoadedData.add(
-                    existingIndex,
-                    readerTagsFeedUiStateMapper.mapErrorTagFeedItem(
-                        tag = tag,
-                        errorType = ErrorType.NoContent,
-                        onTagClick = ::onTagClick,
-                        onRetryClick = ::onRetryClick,
-                    )
-                )
-            }
-        } catch (e: ReaderPostFetchException) {
-            updatedLoadedData.add(
-                existingIndex,
                 readerTagsFeedUiStateMapper.mapErrorTagFeedItem(
                     tag = tag,
-                    errorType = ErrorType.Default,
+                    errorType = ErrorType.NoContent,
                     onTagClick = ::onTagClick,
                     onRetryClick = ::onRetryClick,
+                    onItemEnteredView = ::onItemEnteredView,
                 )
+            }
+        } catch (e: ReaderPostFetchException) {
+            readerTagsFeedUiStateMapper.mapErrorTagFeedItem(
+                tag = tag,
+                errorType = ErrorType.Default,
+                onTagClick = ::onTagClick,
+                onRetryClick = ::onRetryClick,
+                onItemEnteredView = ::onItemEnteredView,
             )
         }
-        _uiStateFlow.update { UiState.Loaded(updatedLoadedData) }
+
+        updateTagFeedItem(updatedItem)
     }
 
-    private fun getUpdatedLoadedData(): MutableList<TagFeedItem> {
+    private fun getLoadedData(uiState: UiState): MutableList<TagFeedItem> {
         val updatedLoadedData = mutableListOf<TagFeedItem>()
-        val currentUiState = _uiStateFlow.value
-        if (currentUiState is UiState.Loaded) {
-            val currentLoadedData = currentUiState.data
-            updatedLoadedData.addAll(currentLoadedData)
+        if (uiState is UiState.Loaded) {
+            updatedLoadedData.addAll(uiState.data)
         }
         return updatedLoadedData
     }
 
+    // Update the UI state for a single feed item, making sure to do it atomically so we don't lose any updates.
+    private fun updateTagFeedItem(updatedItem: TagFeedItem) {
+        _uiStateFlow.update { uiState ->
+            val updatedLoadedData = getLoadedData(uiState)
+
+            // At this point, all tag feed items already exist in the UI.
+            // We need it's index to update it and keep it in place.
+            val existingIndex = updatedLoadedData.indexOfFirst { it.tagChip.tag == updatedItem.tagChip.tag }
```

Comment:
Is it possible to avoid two list operations here (removeAll/add) and just update the element at the selected index?
Something like the following (not tested) code:
```
val existingIndex = updatedLoadedData.indexOfFirst { it.tagChip.tag == updatedItem.tagChip.tag }
if (existingIndex != -1) {
    // Update the item directly at the existing index.
    updatedLoadedData[existingIndex] = updatedItem
}
```

Reply:
Good point! In theory, we shouldn't have duplicate items, so this should work fine. I tested it as well and looks correct. Just sent a new commit!